apis/serving/v1alpha1/servingruntime_webhook.go

```diff
@@ -0,0 +1,189 @@
+/*
+Copyright 2021 IBM Corporation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package v1alpha1
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strconv"
+
+	kservev1alpha "github.com/kserve/kserve/pkg/apis/serving/v1alpha1"
+	"github.com/kserve/kserve/pkg/constants"
+	"github.com/kserve/modelmesh-serving/controllers/autoscaler"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+//+kubebuilder:webhook:path=/validate-serving-modelmesh-io-v1alpha1-servingruntime,mutating=false,failurePolicy=fail,sideEffects=None,groups=serving.kserve.io,resources=servingruntimes;clusterservingruntimes,verbs=create;update,versions=v1alpha1,name=servingruntime.modelmesh-webhook-server.default,admissionReviewVersions=v1
+type ServingRuntimeWebhook struct {
+	Client  client.Client
+	decoder *admission.Decoder
+}
+
+func (s *ServingRuntimeWebhook) Handle(ctx context.Context, req admission.Request) admission.Response {
+	var srAnnotations map[string]string
+
+	if req.Kind.Kind == "ServingRuntime" {
+		servingRuntime := &kservev1alpha.ServingRuntime{}
+		err := s.decoder.Decode(req, servingRuntime)
+		if err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		srAnnotations = servingRuntime.ObjectMeta.Annotations
+	} else {
+		clusterServingRuntime := &kservev1alpha.ClusterServingRuntime{}
+		err := s.decoder.Decode(req, clusterServingRuntime)
+		if err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		srAnnotations = clusterServingRuntime.ObjectMeta.Annotations
+	}
```

Comment:
um I am not sure either. Does the option affect something for webhook validation?  Could you please elaborate more ?

Reply:
Looking at this line: https://github.com/kserve/modelmesh-serving/blob/3f018571f0dd79bb9905dde6855642c73478c048/controllers/servingruntime_controller.go#L265

there are 3 cases when the Deployment is not created (i.e. the ServingRuntime is ignored, because it is not targeted for modelmesh).

However, in the webhook I understand validations are always done, even if the ServingRuntime is not targeted for modelmesh. 

So, I'm wondering if validations should be skipped if the ServingRuntime is not targeted for modelmesh? Skipping the validation would allow the resource to be created with invalid annotations, but since is not targeted for modelmesh, it shouldn't matter. That would give the chance to some other controller (kserve?) to process the ServingRuntime.

Please note that this is more a question. I don't have any problem if validations are performed always :-)

Reply:
I think @israel-hdez's point is a good one!
The validating webhook for the ServingRuntime will be used to validate all created SR resources. If a cluster has both `kserve/kserve` and `kserve/modelmesh-serving` installed, the webhook will be called for SRs that are intended for just `kserve/kserve` (eg. ones without `multiModel: true`). Just in case there ends up being conflicting validations, I think this webhook should skip validation checks for SRs without `multiModel: true` as @israel-hdez suggested.

Reply:
Keep in mind, `kserve/kserve` runtime does not have the annotations because they are using these annotation for inferenceService. So it is only for `modelmesh-serving`. Hence, if there are these annotations in the ServingRuntime, it is for `modelmesh-serving` and it should be validated. 
Regarding `multiModel: true`, the webhook does not check the lifecycle of the deployment. It only validates the annotations such as autoscaler "hpa2" is not allowed or minReplicas -1 is not allowed .
This kind of checking does not hurt for  `multiModel: true` case and it is even better not to create this kind of possible problematic ServingRuntime at the beginning. 

Please let me know your idea. I am open to listen and change the code as well :)

Reply:
Just to clarify any confusion, I'm not suggesting to check the deployment here.

My suggestion is to simply skip validations for SR resources that are not targeted for ModelMesh. In terms of code, I'm simply suggesting to add a check like the follwing _before_ the validations:

```go
if !servingRuntime.Spec.IsMultiModelRuntime() {
    return admission.Allowed("Not validating ServingRuntime because it is not ModelMesh compatible")
}
```

Probably this code will work, but I hope it communicates my idea.

Reply:
Ok I will add the multiMode stuff but I don't know how to find if this is for modelmesh or kserve with ServingRuntime. the deployment mode is added to inference service. 
does it sound good?

Reply:
My understanding is that `multiModel: true` on the SR spec is what distinguishes between a ServingRuntime for ModelMesh or for `kserve/kserve`. I know that the modelmesh controller ignores SR specs without it, but I am not as familiar with the internals of kserve... Does `kserve/kserve` use/reconcile SR specs with `multiModel: true`?

Reply:
Uhm... well... I'm not familiar with modelmesh nor kserve. I'm still learning. I have only read chunks of the code from both codebases. With this limited knowledge, I can only tell:

* I couldn't find a controller on kserve/kserve that would reconcile SR resources.
* There is a controller that reconcilies InferenceService resources. It needs to fetch the related SR or CSR (if any) for the reconciliation.
  * For cases when the runtime is implicit (i.e. no `runtime`) **apparently** it won't pick [C]SR resources that have `multiModel: true`. Code reference: https://github.com/kserve/kserve/blob/e3a10812671f816565120fa19f90ab667a30252b/pkg/apis/serving/v1beta1/predictor_model.go#L111-L126).
  * For cases when the runtime is explicit, it doesn't seem to stop reconciliation even if `multiModel: true` is set. Code reference: https://github.com/kserve/kserve/blob/e3a10812671f816565120fa19f90ab667a30252b/pkg/controller/v1beta1/inferenceservice/components/predictor.go#L93-L131